File: go/classic/hIndex_274.go
package classic

import "sort"

func hIndex(citations []int) int {
	sort.Ints(citations)
	res := 0
	for i, citation := range citations {
		// i 表示的是篇数 因为排序了 所以后面的篇数一定比是现在 i 的引用多的
		// len(citations) - i 后面的大于 当前论文引用数的论文数量
		// 所以后面的论文 至少引用了这么 citation 篇 之前的论文 一定小于 citation 篇
		if len(citations)-i <= citation && len(citations)-i > res {
			res = len(citations) - i
		}
	}
	return res
}

File: java/src/HIndex_274.java
import java.util.Arrays;

public class HIndex_274 {

    // h 表示的 至少有 h 篇论文被引用了 h 次
    public int hIndex(int[] citations) {
        if (citations.length == 0) return 0;
        // 先从小到大排序
        Arrays.sort(citations);

        int res = 0;
        int total = citations.length;

        for (int i = citations.length - 1; i >= 0; i--) {
            // 这个 i 后面又多少个 后面的 一定比前面的多
            int atLeast = total - i;
            int mayBeRes = Math.min(atLeast, citations[i]);
            res = Math.max(res, mayBeRes);
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println(new HIndex_274().hIndex(new int[]{3,0,6,1,5}));
    }
}
